### PR TITLE
Emissions

### PR DESCRIPTION
**Pull request by talekhinezh**
* Implement emissions for each proposal submitted by proposer

### CODE REVIEW

**Review by siy — APPROVED** [review]
LGTM

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/StakedValidators.java [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+    private final int maxValidators;
+
+    private StakedValidators(
+        int minValidators,
+        int maxValidators,
+        Set<ValidatorParticle> validatorParticles,
+        Map<ECPublicKey, UInt256> stakedAmounts
+    ) {
+    	this.minValidators = minValidators;
+    	this.maxValidators = maxValidators;
```
Comment (LukasGasior1) [inline radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/StakedValidators.java:63]: spaces -> tabs

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/StakedValidatorsReducer.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+import java.util.function.BiFunction;
+import java.util.function.Supplier;
+
+/**
+ * Reduces particles to Registered Validators
+ */
+public final class StakedValidatorsReducer implements StateReducer<StakedValidators> {
+
+    private final int minValidators;
+    private final int maxValidators;
```
Comment (LukasGasior1) [inline radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/StakedValidatorsReducer.java:37]: spaces -> tabs

**Review by michalzajda — APPROVED** [review]

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/RadixEngineModule.java [hunk 3]
```diff
@@ -104,8 +97,9 @@ PostParsedChecker checker(Set<PostParsedChecker> checkers) {
 		//   .toWindowedSet(initialValidatorSet, RegisteredValidatorParticle.class, p -> p.getAddress(), 2)
 		//   .build();
 
-		radixEngine.addStateReducer(new ValidatorsReducer(), true);
-		radixEngine.addStateReducer(new StakesReducer(), true);
+		radixEngine.addStateReducer(stakedValidatorsReducer, true);
+		//radixEngine.addStateReducer(new DeprecatedStakesReducer(), true);
+		radixEngine.addStateReducer(new InflationReducer(), "inflation", true);
```
Comment (michalzajda) [inline radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/RadixEngineModule.java:102]: So, we can say that Reducers are "configuration" for ApplicationStateComputer?

We keep all of them in the RadixEngine. Providing the name is defined, it's going to be the key to get it from the map.

Depending on particle spin we use output or input function defined in the Reducer.

The processCheckSpin is not only interpreting the spin, but actually provides execution context for any function.

Comment (talekhinezh) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/RadixEngineModule.java:102]: I think sounds about right. Reducers map groups of substate to higher level state constructs, `stakedValidatorsReducer` being the most important as it maps registered validators and stake to `BFTValidatorSet` which is used for consensus. `InflationReducer` isn't really used much except for testing.

**Pull request merged**